### PR TITLE
fix(listing): pick exact or create new one on update

### PR DESCRIPTION
**Pull request by shcheklein**
Partially addresses and stabilizes https://github.com/iterative/datachain/issues/725

Fixes a few situations in the way we find and suggest an existing cached listing when we do an `update=True`.

## 1. There was a bug in the origin code:

```python
# choosing the smallest possible one to minimize update time
listing = sorted(listings, key=lambda ls: len(ls.name))[0]
```

This code actually was picking the "biggest" one instead. There should have been `reverse=True`.

So, in a situation like:

`file:///whatever/dir1` exists and cached
`file:///whatever/` exists and cached

and we were trying to update the `file:///whatever/dir1` - it was updating the `file:///whatever/` instead.

## 2. Then, in a situation like:

`file:///whatever/` exists and cached
`file:///whatever/dir1` doesn't yet exist

and we do "update=True" for the `file:///whatever/dir1`

it was updating the whole `file:///whatever/` which can be arbitrary large. And we don't want to do this.

## So, bottom line is:

 Generalizing a bit code for both of those cases led to the code in this PR.

### CODE REVIEW

File: tests/func/test_catalog.py [hunk 1]
```diff
@@ -840,13 +840,17 @@ def test_listing_stats(cloud_test_catalog):
 
     catalog.enlist_source(f"{src_uri}/dogs/", update=True)
     stats = listing_stats(src_uri, catalog)
+    assert stats.num_objects == 7
```
Comment (shcheklein) [inline tests/func/test_catalog.py:843]: [C]: not sure why we expected here any change to the `src_uri` listing, after listing the subset. It seems it is better to create a second listing for `/dogs` in this case (or at least update the whole `src_uri`).

Comment (shcheklein) [reply tests/func/test_catalog.py:843]: cc @ilongin you might know this better

File: src/datachain/lib/dc.py [hunk 2]
```diff
@@ -430,7 +430,11 @@ def parse_uri(
         if listings:
             if update:
                 # choosing the smallest possible one to minimize update time
-                listing = sorted(listings, key=lambda ls: len(ls.name))[0]
+                listing = sorted(listings, key=lambda ls: len(ls.name), reverse=True)[0]
```
Comment (shcheklein) [inline src/datachain/lib/dc.py:433]: @ilongin why `parse_uri` is exposed externally? also it's a bit of a strange name tbh. Could you please give more context?

Comment (ilongin) [reply src/datachain/lib/dc.py:433]: Yes, this should not be exposed. Probably this can be removed from `DataChain` completely in the first place ...

Comment (shcheklein) [reply src/datachain/lib/dc.py:433]: Kk, I'll move it

File: tests/func/test_catalog.py (hunk 1 above)
Comment (ilongin) [reply tests/func/test_catalog.py:843]: I think you are right and this was a mistake / bug in the listing. Creating new separate `/dogs` seems like correct thing to do here.

File: src/datachain/lib/dc.py [hunk 3]
```diff
@@ -427,17 +427,18 @@ def parse_uri(
...
+        # if no need to update - choosing the most recent one;
+        # otherwise, we'll using the exact original `ds_name`` in this case:
+        # - if a "bigger" listing exists, we don't want to update it, it's better
+        #   to create a new "smaller" one on "update=True"
+        # - if an exact listing exists it will have the same name as `ds_name`
+        #   anyway below
+        if listings and not update:
+            listing = sorted(listings, key=lambda ls: ls.created_at)[-1]
+
+        # For local file system we need to fix listing path / prefix
```
Comment (shcheklein) [inline src/datachain/lib/dc.py:439]: @ilongin Q: why do we need to do this only for the local file system?

File: tests/unit/test_listing.py [hunk 4]
```diff
@@ -38,20 +39,43 @@ def _tree_to_entries(tree: dict, path=""):
 @pytest.fixture
 def listing(test_session):
     catalog = test_session.catalog
-    dataset_name, _, _, _ = DataChain.parse_uri("s3://whatever", test_session)
+    dataset_name, _, _, _ = DataChain.parse_uri("file:///whatever", test_session)
```
Comment (shcheklein) [inline tests/unit/test_listing.py:42]: [C]: minor semi-related change - there is not reason to use `s3`, tests were failing since it was hitting actual token update w/o proper fixtures. cc @ilongin

**Review by dreadatour — APPROVED** [review]
👍

**Review by amritghimire — APPROVED** [review]
